### PR TITLE
rinv command get error message on Habanero

### PR DESCRIPTION
**Pull request by cxhong**
Resolves Issue #2335  

For Hananerro system,  it uses fru id 2 to get ```model``` subcommand  and  fru id 43 to get ``firm`` subcommand.

Test results:
````
# rinv c712f11n06 model
c712f11n06: Backplane Chassis Part Number: 8348-21C
# rinv c712f11n06 firm
c712f11n06: BMC Firmware: 2.08
c712f11n06: System Firmware Product Version: IBM-habanero-ibm-OP8_v1.6_0.37
c712f11n06: System Firmware Product Additional Info: hostboot-bc98d0b-3385c16
c712f11n06: System Firmware Product Additional Info: occ-0362706
c712f11n06: System Firmware Product Additional Info: skiboot-5.1.5-9f8b09a
c712f11n06: System Firmware Product Additional Info: hostboot-binaries-43d5a59
c712f11n06: System Firmware Product Additional Info: habanero-xml-a71550e-209a017
c712f11n06: System Firmware Product Additional Info: capp-ucode-105cb8f
````

### CODE REVIEW

**Comment by whowutwut** [comment]
@cxhong  Do all the following commands work for Habanero, or only address the problem mentioned in the issue? 

```
   OpenPOWER server specific:
        rinv noderange {model|serial|deviceid|uuid|guid|vpd|mprom|firm|all}
```

**Comment by cxhong** [comment]
ran all the above subcomands, looks like they are all working.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 1]
```diff
@@ -8459,6 +8466,36 @@ sub genhwtree
 
 }
 
+##########################################################################
+# To check if this is Habanero system
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: I think we should change "Habanero" to say something different, I suspect this is consistent with backplane boards of the same MFG.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 2]
```diff
@@ -3299,7 +3302,11 @@ sub initfru_zero {
 
         if ($sessdata->{skipotherfru} and isopenpower($sessdata)) {
             # For openpower servers, fru 3 is used to get MTM/Serial information, fru 47 is used to get firmware information
-            @{$sessdata->{frus_for_openpower}} = qw(3 47);
+            if (isHabanero($sessdata)) {
+                @{$sessdata->{frus_for_openpower}} = qw(2 43);
+            } else {
+                @{$sessdata->{frus_for_openpower}} = qw(3 47);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: I think here we should put the comment: 

```
# For OpenPower for HPC, fru 3 = mtm/serial, fru 47 = firmware info
```

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 3]
```diff
@@ -3299,7 +3302,11 @@ sub initfru_zero {
 
         if ($sessdata->{skipotherfru} and isopenpower($sessdata)) {
             # For openpower servers, fru 3 is used to get MTM/Serial information, fru 47 is used to get firmware information
-            @{$sessdata->{frus_for_openpower}} = qw(3 47);
+            if (isHabanero($sessdata)) {
+                @{$sessdata->{frus_for_openpower}} = qw(2 43);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Here we should put the comment: 

```
# For OpenPower for Big Data machines, fru 2 = mtm/serial, fru 43 = firmware info
```

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 4]
```diff
@@ -3299,7 +3302,11 @@ sub initfru_zero {
 
         if ($sessdata->{skipotherfru} and isopenpower($sessdata)) {
             # For openpower servers, fru 3 is used to get MTM/Serial information, fru 47 is used to get firmware information
-            @{$sessdata->{frus_for_openpower}} = qw(3 47);
+            if (isHabanero($sessdata)) {
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Let's remove this comment above and replace with suggested comments below.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 5]
```diff
@@ -8459,6 +8466,36 @@ sub genhwtree
...
 
+##########################################################################
+# To check if this is Habanero system
+# we identified Hananero via Chassis Part number : 8348-21C
+##########################################################################
+sub isHabanero
+{
+    my $sessdata = shift;
+
+    my $bmc_addr     = $sessdata->{ipmisession}->{bmc};
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Seems like we could roll this pre_cmd into a function, but maybe not necessary for this pull request. ...  not sure how much other places in the code is doing the same thing.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 6]
```diff
@@ -8459,6 +8466,36 @@ sub genhwtree
 
 }
 
+##########################################################################
+# To check if this is Habanero system
+# we identified Hananero via Chassis Part number : 8348-21C
+##########################################################################
+sub isHabanero
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Could we change this function name to `isopenpower_BD`?   Habanero is a code word for a box, and so I think this code is applicable to other machines ...

File: xCAT-server/lib/xcat/plugins/ipmi.pm (hunk 4 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: ok

File: xCAT-server/lib/xcat/plugins/ipmi.pm (hunk 3 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: ok

File: xCAT-server/lib/xcat/plugins/ipmi.pm (hunk 5 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: I will not change here for now

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 7]
```diff
@@ -3288,7 +3288,7 @@ sub initfru_zero {
         if ($_->{encoding} == 3) {
             $fru->value($_->{value});
         } else {
-            next;
+
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Is this delete of line intentional?

Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: how that happened :(

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 8]
```diff
@@ -8459,6 +8467,36 @@ sub genhwtree
 
 }
 
+##########################################################################
+# To check if this is openpower Big Data system
+# we identified it via Chassis Part number : 8348-21C
+##########################################################################
+sub isopenpower_bd
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: Hi, @cxhong , I suggest we don't call "ipmitool" in ipmi.pm to judge it is a Habanero server.
The current logic  is, we use ipmitool fru 2 to check the mtm of the server, then if it is a habanero, we call raw data to output of information of fru 2. 
Very strange, right? Use ipmitool check the MTM first, then use raw data output its MTM.

File: xCAT-server/lib/xcat/plugins/ipmi.pm [hunk 9]
```diff
@@ -3298,8 +3301,13 @@ sub initfru_zero {
...
 
         if ($sessdata->{skipotherfru} and isopenpower($sessdata)) {
-            # For openpower servers, fru 3 is used to get MTM/Serial information, fru 47 is used to get firmware information
-            @{$sessdata->{frus_for_openpower}} = qw(3 47);
+            # For openpower Big Data servers, fru 2 has MTM/Serial and fru 43 has firmware information
+            if (isopenpower_bd($sessdata)) {
+                @{$sessdata->{frus_for_openpower}} = qw(2 43);
+            # For openpower HPC servers, fru 3 has MTM/Serial and fru 47 has firmware information
+            } else {
+                @{$sessdata->{frus_for_openpower}} = qw(3 47);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: We can use 2,3,43,47 to cover both Habanero and Firestone/Minsky servers.
Just use the pattern in line 2673 to get related MTMS for each server.

Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: thanks for this suggestion.  It worked for Firestone servers.  but need to modify one of checking for habanero servers.

File: xCAT-server/lib/xcat/plugins/ipmi.pm (hunk 8 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/ipmi.pm]: yeh,  we can remove this checking and use 2,3,43,47 for ```{frus_for_openpower}```

**Comment by zet809** [comment]
Hi, @cxhong , the pull request looks good to me, will you pls answer the 2 question above before I merge the code?

**Comment by cxhong** [comment]
@zet809 , for the committed code, I didn't call "ipmitool" in the ipmi.pm.  I used  2,3,43,47 to cover both Habanero and Firestone/Minsky servers as you suggested.  what other questions you asked above?

**Comment by zet809** [comment]
Thanks very much, @cxhong , looks good to me, I agree to merge.

**Pull request merged**